Add tests for more properties

niccokunzmann opened the pull request:
This PR fixes #655, making icalendar compatible to RFC 7986.

todo:

- all properties
  - [x] LAST-MODIFIED Property
  - [x] [5.5](https://www.rfc-editor.org/rfc/rfc7986.html#section-5.5). URL Property
  - [x] [5.7](https://www.rfc-editor.org/rfc/rfc7986.html#section-5.7). REFRESH-INTERVAL Property
  - [x] [5.8](https://www.rfc-editor.org/rfc/rfc7986.html#section-5.8). SOURCE Property
  - [x] [5.10](https://www.rfc-editor.org/rfc/rfc7986.html#section-5.10). IMAGE Property
  - [x] [5.11](https://www.rfc-editor.org/rfc/rfc7986.html#section-5.11). CONFERENCE Property
- [x] proper value types
- [x] add label of compatibility to docs

https://icalendar--877.org.readthedocs.build/en/877/

SashankBhamidi commented on src/icalendar/tests/test_conference.py:
```diff
@@ -0,0 +1,144 @@
+    return component_class()
+
+
+def test_no_conferenes(component):
```
Valid and should be implemented.

SashankBhamidi commented on src/icalendar/prop/__init__.py:
```diff
@@ -77,10 +76,13 @@ class vBinary:
+    obj: str
 
-    def __init__(self, obj):
+    def __init__(self, obj, params: dict[str:str] | None = None):
```
Valid and should be implemented.

SashankBhamidi commented on src/icalendar/tests/test_rfc_7986.py:
```diff
@@ -212,3 +211,56 @@ def test_set_if_multiple_colors(
+
+def test_refresh_interval_default(calendar: Calendar):
+    """REFRESH-INTERVAL default."""
+    assert calendar.get("REFESH-INTERVAL") is None
```
Valid and should be implemented.

niccokunzmann commented on src/icalendar/prop/conference.py:
```diff
@@ -0,0 +1,96 @@
+    language: list[str] | str | None = None
+
+    @classmethod
+    def from_uri(cls, uri: vUri):
+        """Create a Conference from a URI."""
+        return cls(
+            uri,
+            feature=uri.params.get("feature"),
+            label=uri.params.get("label"),
+            language=uri.params.get("language"),
+        )
```
51bcca7

niccokunzmann commented:
@SashankBhamidi Thanks for your review! I addressed the issues, I believe. Let me know if there is anything else to do and feel free to merge it if you are satisfied.

stevepiercy commented on src/icalendar/attr.py:
```diff
@@ -1763,13 +1829,157 @@ def set_end_with_locking(
+        iCalendar object or in "VEVENT", "VTODO", or "VJOURNAL" calendar
+        components.
+
+    .. note::
+
+        At the present moment, this property is read-only. If you require a setter,
+        please open an issue or a pull request.
```
Based on this note, I think you expect this method and its docstring to be published. However when I alter the docs locally, it still doesn't display. Does it need to be added to `__all__` or something else?

```rst
.. automodule:: icalendar.attr
   :members:
   :private-members:
```

Perhaps these will help?

- https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-option-automodule-private-members
- https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoproperty

Alternatively you could follow the pattern for `icalendar.attr.property_del_duration`, `icalendar.attr.property_get_duration`, and `icalendar.attr.property_set_duration`.

stevepiercy replied:
Ah, OK, I expected to see its private name of `_get_images`. I didn't understand where it should appear. Noted. Thank you!

stevepiercy commented on src/icalendar/attr.py:
```diff
@@ -1763,13 +1829,157 @@ def set_end_with_locking(
+images_property = property(_get_images)
+
+
+def _get_conferences(self: Component) -> list[Conference]:
```
Also private member, not displayed in docs. Do you want it to display?

niccokunzmann replied:
See https://github.com/collective/icalendar/pull/877#discussion_r2387799765

stevepiercy reviewed (changes requested):
@niccokunzmann a couple of suggestions and questions.

I like @SashankBhamidi suggestions, too.

niccokunzmann commented on src/icalendar/tests/test_rfc_7986.py:
```diff
@@ -212,3 +211,84 @@ def test_set_if_multiple_colors(
+        calendar.refresh_interval = timedelta(seconds=-1)
+
+
+def test_0_refresh_interval(calendar: Calendar):
+    """REFRESH-INTERVAL zero."""
+    calendar.refresh_interval = timedelta(0)
+    assert calendar.refresh_interval == timedelta(0)
+    assert calendar["REFRESH-INTERVAL"].dt == timedelta(0)
```
deleted the test, this is now covered by test_invalid_refresh_interval

niccokunzmann commented on src/icalendar/prop/conference.py:
```diff
@@ -0,0 +1,98 @@
+    Example:
+        The following are examples of this property:
+
+        .. code-block:: ical
```
Ok! Hm... AI hallucination :)

SashankBhamidi replied:
> Ok! Hm... AI hallucination :)

What do you use?

niccokunzmann replied:
Windsurf AI and then probably Visual Studio Code does its thing, too.

stevepiercy commented on src/icalendar/attr.py:
```diff
@@ -1763,13 +1829,157 @@ def set_end_with_locking(
         raise ValueError(msg)
 
 
+def _get_images(self: Component) -> list[Image]:
```
Removed display of private methods.

SashankBhamidi commented:
Great work! Feel free to merge after the doc fix.

niccokunzmann commented on src/icalendar/prop/conference.py:
```diff
@@ -0,0 +1,98 @@
+    def to_uri(self) -> vUri:
+        """Convert the Conference to a vUri."""
+        params = {}
+        if self.feature:
+            params["FEATURE"] = self.feature
+        if self.label:
+            params["LABEL"] = self.label
+        if self.language:
+            params["LANGUAGE"] = self.language
```
This goes into a new issue because it needs to be tested. Could you open this after merge?
